src/bitcoind.cpp

```diff
@@ -105,24 +187,34 @@ static bool AppInit(int argc, char* argv[])
             // InitError will have been called with detailed error, which ends up on console
             return false;
         }
-        if (args.GetBoolArg("-daemon", false)) {
-#if HAVE_DECL_DAEMON
-#if defined(MAC_OSX)
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
-#endif
+        if (args.GetBoolArg("-daemon", DEFAULT_DAEMON) || args.GetBoolArg("-daemonwait", DEFAULT_DAEMONWAIT)) {
+#if HAVE_DECL_FORK
             tfm::format(std::cout, PACKAGE_NAME " starting\n");
 
             // Daemonize
-            if (daemon(1, 0)) { // don't chdir (1), do close FDs (0)
-                return InitError(Untranslated(strprintf("daemon() failed: %s\n", strerror(errno))));
+            switch (fork_daemon(1, 0, daemon_ep)) { // don't chdir (1), do close FDs (0)
+                case 0: // Child: continue.
+                    // If -daemonwait is not enabled, immediately send a success token the parent.
+                    if (!args.GetBoolArg("-daemonwait", DEFAULT_DAEMONWAIT)) {
+                        daemon_ep.TokenWrite(1);
+                        daemon_ep.Close();
+                    }
+                    break;
+                case -1: // Error happened.
+                    return InitError(Untranslated(strprintf("fork_daemon() failed: %s\n", strerror(errno))));
+                default: { // Parent: wait and exit.
+                        int token = daemon_ep.TokenRead();
+                        if (token) { // Success
+                            exit(EXIT_SUCCESS);
+                        } else { // fRet = false or token read error (premature exit).
+                            tfm::format(std::cout, "Error during initializaton - check debug.log for details\n");
+                            exit(EXIT_FAILURE);
+                        }
+                    }
```

Comment:
indentation-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
            switch (fork_daemon(1, 0, daemon_ep)) { // don't chdir (1), do close FDs (0)
            case 0: // Child: continue.
                // If -daemonwait is not enabled, immediately send a success token the parent.
                if (!args.GetBoolArg("-daemonwait", DEFAULT_DAEMONWAIT)) {
                    daemon_ep.TokenWrite(1);
                    daemon_ep.Close();
                }
                break;
            case -1: // Error happened.
                return InitError(Untranslated(strprintf("fork_daemon() failed: %s\n", strerror(errno))));
            default: // Parent: wait and exit.
                int token = daemon_ep.TokenRead();
                if (token) { // Success
                    exit(EXIT_SUCCESS);
                } else { // fRet = false or token read error (premature exit).
                    tfm::format(std::cout, "Error during initializaton - check debug.log for details\n");
                    exit(EXIT_FAILURE);
               }
```